lib/kubernetes-deploy/resource_cache.rb

```diff
@@ -0,0 +1,66 @@
+# frozen_string_literal: true
+module KubernetesDeploy
+  class ResourceCache
+    def initialize(namespace, context, logger)
+      @namespace = namespace
+      @context = context
+      @logger = logger
+
+      @mutexes = Hash.new { |hash, key| hash[key] = Mutex.new }
+      @data = {}
+    end
+
+    def get_instance(kind, resource_name, raise_if_not_found: false)
+      instance = use_or_populate_cache(kind).fetch(resource_name, {})
+      if instance.blank? && raise_if_not_found
+        raise KubernetesDeploy::Kubectl::ResourceNotFoundError, "Resource does not exist (used cache for kind #{kind})"
+      end
+      instance
+    rescue KubectlError
+      {}
+    end
+
+    def get_all(kind, selector = nil)
+      instances = use_or_populate_cache(kind).values
+      return instances unless selector
+
+      instances.select do |r|
+        labels = r.dig("metadata", "labels") || {}
+        labels >= selector
+      end
+    rescue KubectlError
+      []
+    end
+
+    private
+
+    def statsd_tags
+      { namespace: @namespace, context: @context }
+    end
+
+    def use_or_populate_cache(kind)
+      @mutexes[kind].synchronize do
```

Comment:
Is there a race condition in creating the mutex?

Reply:
Hm, yeah, probably a really really tiny one? Not sure how I'd write tests for that, or whether it's likely enough to warrant handling with another mutex. 😕

Reply:
I actually think its worth protecting against it. But agree its probably too hard to capture with a test.

Reply:
Yea, Danny is right I think. There's actually two potential problems here as far as I can tell:
* Two threads with different `kind` values might access the `@mutexes` hash in parallel and hash access in Ruby isn't thread-safe.
* Two threads with the same `kind` values might race and both trigger the creation of a new mutex and one thread's mutex can overwrite the other thread's mutex.

Would it make sense to have one ResourceCache instance per kind? Rather than one instance that's shared by all kinds. That would simplify this mutex stuff a lot I think.

Reply:
> Would it make sense to have one ResourceCache instance per kind? Rather than one instance that's shared by all kinds. That would simplify this mutex stuff a lot I think.

I'm having trouble imagining how that would work. The list of kinds changes dynamically on every polling cycle, so we'd still need a higher-level thing generating/managing the mutexes. Even the list of possible kinds can't be anticipated, because we don't explicitly support all of the possibilities (and custom resources are a thing, though we aren't handling them dynamically _yet_).